Models and admin integration

e-kolpakov opened the pull request:
**Description:** This PR adds models and admin integration to `enterprise` app.

**JIRA:** [ENT-33](https://openedx.atlassian.net/browse/ENT-33) and [ENT-21](https://openedx.atlassian.net/browse/ENT-21)

**Dependencies:** Cross-dependent on https://github.com/edx/edx-platform/pull/13723

**Sandbox:**
* LMS: http://pr13723.sandbox.opencraft.hosting/
* Studio: http://studio-pr13723.sandbox.opencraft.hosting/

**Testing instructions:**
1. This PR contains only Django admin integration so far - smoke testing Django admin using Enterprise Customer should be enough.
2. One extension to default Django admin features is ability to export list of Enterprise Customers to CSV.
2.1. Open "Enterprise Customer" page in Django admin
2.2. Select all/some Enterprise Customers.
2.3. In "Actions" dropdown select "Export to CSV", than press "Go" button. _Expected result:_ selected Enterprise Customers are exported to CSV

**Reviewers:**
- [x] @mtyaka 
- [ ] @mattdrayer 
- [ ] @edx/ecommerce

clintonb commented:
It looks like you're making a new Django app. @jmbowman created a cookiecutter template (https://github.com/edx/cookiecutter-django-app) for this purpose. The template is based on our most recent apps, and incorporates our most up-to-date knowledge of file layout, testing, etc. It would be great if that were used to start this project so that we are all on the same page.

e-kolpakov commented:
@clintonb Thanks, I was not aware of that cookiecutter app, but I'll take a look. However, I've based it on edx-organizations app, which looks like it was created using cookiecutter. So I woul expect to not find any significant differences.

clintonb commented:
edx/edx-organizations existed well before the cookiecutter template, which was merged less than two months ago.

e-kolpakov commented:
@clintonb ok, I'll take a look anyway, thank you for the hint.

mtyaka commented on enterprise/admin.py:
```diff
@@ -0,0 +1,85 @@
+    """
+
+    list_display = ("name", "origin", "active", "short_code", "data_sharing_policy", "logo", "created", "modified")
+    search_fields = ("name", "short_code", "data_sharing_policy", "created", "modified")
```
I think we can hide "modified" since the models are immutable and "modified" should always equal "created".

mtyaka commented on enterprise/models.py:
```diff
@@ -2,42 +2,238 @@
+        populate_from="name", unique=True, editable=False,
+        help_text=_("Short code will be used to reference this Enterprise Customer in other parts of the system,"
+                    "i.e. SSO, ecommerce app, analytics etc.")
+    )
```
nit: When adding a new EnterpriseCustomer it currently isn't possible to override the default value for `short_code`, it always generates it from the name. It would be nice if it was possible to provide a custom value for `short_code`.

mtyaka commented on enterprise/models.py:
```diff
@@ -2,42 +2,238 @@
+                                         requested data sharing (learners are allowed to opt-in) or required data
+                                         sharing (learners must share their data)
+        sso_providers (CharField): list of comma-separated SSO provider slugs. Providers in this list should link
+                                    users registered with that provider with Enterprise Customer
```
The two lines above should be removed since the `sso_provider` field no longer exists.

mtyaka commented on enterprise/models.py:
```diff
@@ -2,42 +2,238 @@
+                                         requested data sharing (learners are allowed to opt-in) or required data
+                                         sharing (learners must share their data)
+        sso_providers (CharField): list of comma-separated SSO provider slugs. Providers in this list should link
+                                    users registered with that provider with Enterprise Customer
```
The `sso_providers` field no longer exists.

mtyaka commented on enterprise/admin.py:
```diff
@@ -0,0 +1,85 @@
+    search_fields = ("name", "short_code", "data_sharing_policy", "created", "modified")
+    exclude = ("members",)
+
+    ordering = ["-created"]
```
nit: Since we're ordering by `created` it would be nice if the field was indexed. I'm not sure how easy it is to do that with a `TimeStampedModel`.

mtyaka commented on enterprise/models.py:
```diff
@@ -2,42 +2,238 @@
+    """
+
+    enterprise_customer = models.ForeignKey(EnterpriseCustomer, blank=False, null=False)
+    user_id = models.IntegerField(null=False, blank=False)
```
Why is this not a `ForeignKey` (genuine question)?

e-kolpakov replied:
@mtyaka because we don't want to tie enterprise to anything in LMS, as there are plans to make it a separate service/app in the future. We've discussed that with @mattdrayer on HipChat on Friday and decided that storing user_ids instead of foreign key will allow for greater flexibility in future, while also not significantly complicating current setup.

mtyaka replied:
Makes sense, thanks.

clintonb replied:
`PositiveIntegerField`

mtyaka commented on tests/test_models.py:
```diff
@@ -6,22 +6,77 @@
+        customer = EnterpriseCustomerFactory(id=150, name="QWERTY")
+        expected_to_str = "<EnterpriseCustomer 150: QWERTY>"
+        self.assertEqual(method(customer), expected_to_str)
 
```
nit: Consider adding a test that makes sure a new `EntepriseCustomerHistory` instance is created every time a new `EnterpriseCustomer` model is created or updated.

mtyaka commented on enterprise/admin.py:
```diff
@@ -0,0 +1,85 @@
+        """
+        Don't allow deletes.
+        """
+        return False
```
Should we add `readonly_fields = EnterpriseCustomerHistory._meta.get_all_field_names()` to make the model completely readonly?

mtyaka commented:
:+1:

- [x] I tested this: Creation of new models in the django admin, CSV export action.
- [x] I read through the code
- [x] Includes documentation

e-kolpakov commented:
Rebased from d6b98c6b446a6c4b8a5aa859c3d3e939e3f8a40b

e-kolpakov commented:
@edx/ecommerce please take a look - this PR introduces Enterprise Customer entity and `enterprise` app.

mattdrayer commented on enterprise/admin.py:
```diff
@@ -0,0 +1,110 @@
+
+    class Meta(object):
+        """
+        Meta class for EnterpriseCustomerHistory admin model.
```
Update docstring

mattdrayer commented on enterprise/admin.py:
```diff
@@ -0,0 +1,110 @@
+        Make all fields readonly when editing existing model.
+        """
+        if obj:  # editing an existing object
+            return 'enterprise_customer', 'user_id'
```
Is this `enterprise_customer` or `enterprise_customer_id`?

mattdrayer commented on enterprise/models.py:
```diff
@@ -2,42 +2,235 @@
+
+
+@python_2_unicode_compatible
+class EnterpriseCustomerMember(models.Model):
```
Unfortunately "member" is an overloaded term at edX, and it will be difficult to change the existing connotation.  So this model should probably be renamed to `EnterpriseCustomerUser`, which is a better description of what the data actually is, anyway.

mattdrayer replied:
This could be a TimeStampedModel as well

mattdrayer commented on enterprise/models.py:
```diff
@@ -2,42 +2,235 @@
+        help_text=_("Please add only .PNG files for logo images."),
+        null=True, blank=True, max_length=255
+    )
+    data_sharing_policy = models.CharField(
```
I think we need to remove `data_sharing_policy` and related constructs for the present time, until we have figured out how all of the sharing and consent stuff is going to work.

clintonb commented on enterprise/models.py:
```diff
@@ -2,42 +2,235 @@
+                    "i.e. SSO, ecommerce app, analytics etc. It is allowed to override it at creation, but it should "
+                    "never be edited on existing records, as it might break integration with other apps.")
+    )
+    logo = models.ImageField(
```
Use [`django-stdimage`](https://github.com/codingjoe/django-stdimage) to manage image manipulation (e.g. resizing). You should also include some form of validation of file size, resolution, and file format. Help text is easily ignored.

clintonb replied:
Also, please work with @edx/devops to get these uploaded to S3.

e-kolpakov replied:
@clintonb I believe this should be delayed until later - there are no product/functional requirements regarding logo at all (at least I haven't seen them). So I would suggest it as is (or just remove the field altogether) until we've got more details about how co-branding should happen.

mattdrayer commented on test_utils/__init__.py:
```diff
@@ -0,0 +1,6 @@
+"""
+# Since py.test discourages putting __init__.py into test directory (i.e. making tests a package)
+# one cannot import from anywhere under tests folder. However, some utility classes/methods might be useful
+# in multiple test modules. So this package the place to put them.
```
This seems like a good idea to me -- should we consider adding it to the cookiecutter spec?

e-kolpakov replied:
@mattdrayer maybe - I'll see if I could fit it into my week.

clintonb commented on enterprise/models.py:
```diff
@@ -2,42 +2,235 @@
+    objects = EnterpriseCustomerManager()
+    all_objects = models.Manager()
+
+    name = models.CharField(max_length=500, blank=False, null=False, help_text=_("Enterprise Customer name."))
```
The maximum length for `CharField` in MySQL is 255 characters. What's the purpose of this field. Why would it ever need 500 characters? How will the data be displayed/presented?

clintonb commented on enterprise/models.py:
```diff
@@ -2,42 +2,235 @@
+        Meta class for EnterpriseCustomer model.
+        """
+
+        app_label = "enterprise"
```
This is not necessary.

e-kolpakov replied:
@clintonb ... but might still be good for consistence.

clintonb replied:
Consistency with what? If you ever change the name of the app, you now have to change this value everywhere. It's not a huge task; but, why take on the extra work if you don't need to? This is a nit-pick, so do what you wish.

clintonb commented on enterprise/models.py:
```diff
@@ -2,42 +2,235 @@
+        """
+
+        app_label = "enterprise"
+        verbose_name = "Enterprise Customer"
```
Mark both of these for translation.

mattdrayer commented:
@e-kolpakov, I've made a pass through this change set -- main items are changing from "member" to "user", and removing the data sharing bits until we figure all that out.

clintonb commented on enterprise/models.py:
```diff
@@ -2,42 +2,235 @@
+    all_objects = models.Manager()
+
+    name = models.CharField(max_length=500, blank=False, null=False, help_text=_("Enterprise Customer name."))
+    short_code = AutoSlugField(
```
If this is going to be a unique identifier across multiple systems, it should be a non-editable UUID. We have had much success using UUIDs to synchronize data between the marketing site, Programs API and the Catalog Service.

e-kolpakov replied:
@clintonb it is unique, non-editable after a record is created and have a sensible default so that operator don't have to manually enter it.

As an added benefit, it is more human-friendly (you can easily go from "ministry-of-defence" to Ministry of Defence record, but not quite easily from "05aa1b9d-33f6-4e92-a3c5-a6c2d99f0999" to whatever else) and allows operator to override it if required.

clintonb commented on enterprise/models.py:
```diff
@@ -2,42 +2,235 @@
+
+    name = models.CharField(max_length=500, blank=False, null=False, help_text=_("Enterprise Customer name."))
+    short_code = AutoSlugField(
+        max_length=120, blank=True, null=False, editable=True,
```
1. 120 characters is not very short.
2. If you choose to ignore the note about UUIDs, this field should neither allow blank input, be nullable, nor be editable.

e-kolpakov replied:
@clintonb This is all carefully selected:

* 120 chars - not very big as well, but allows to put some sensible slugs for longer names.
* blank=True allows passing empty value from the form, so default value is applied (deep in `AutoSlugField)
* null=False - obviously, we want that field to be always present
* editable=True - allows editing the field, but it is overridden in django admin to disallow edits on existing fields. As far as [documentation](https://docs.djangoproject.com/en/1.10/ref/models/fields/#editable) tells, `editable` only have effect on admin interface.

clintonb commented on enterprise/models.py:
```diff
@@ -2,42 +2,235 @@
 
 
 @python_2_unicode_compatible
 class EnterpriseCustomerHistory(TimeStampedModel):
```
I see no benefit to recreating the wheel here. Use [`django-simple-history`](https://django-simple-history.readthedocs.io/en/latest/).

clintonb commented on enterprise/models.py:
```diff
@@ -2,42 +2,235 @@
-        # TODO: return a string appropriate for the data fields
-        return '<EnterpriseCustomerHistory, ID: {}>'.format(self.id)
+
+        app_label = "enterprise"
```
This is not necessary.

clintonb commented on enterprise/models.py:
```diff
@@ -2,42 +2,235 @@
-        return '<EnterpriseCustomerHistory, ID: {}>'.format(self.id)
+
+        app_label = "enterprise"
+        verbose_name = "Enterprise Customer Member"
```
Mark these for translation.

clintonb commented on setup.py:
```diff
@@ -45,7 +45,13 @@ def get_version(*file_paths):
     include_package_data=True,
-    install_requires=["django-model-utils>=2.0"],
+    install_requires=[
+        "django-model-utils>=2.0",
```
Django should also be a ranged requirement (e.g. `Django>=1.8.15,<1.11`).

clintonb commented on test_utils/factories.py:
```diff
@@ -0,0 +1,65 @@
+
+        model = EnterpriseCustomer
+
+    name = "Enterprise Customer 1"
```
Use `FuzzyText`.

clintonb commented on test_utils/factories.py:
```diff
@@ -0,0 +1,65 @@
+
+        model = EnterpriseCustomerHistory
+
+    name = "Enterprise Customer 1"
```
Use `FuzzyText`.

clintonb commented on test_utils/factories.py:
```diff
@@ -0,0 +1,65 @@
+        model = EnterpriseCustomerMember
+
+    enterprise_customer = factory.SubFactory(EnterpriseCustomerFactory)
+    user_id = 1
```
Use `FuzzyInteger`.

clintonb commented on tests/test_actions.py:
```diff
@@ -0,0 +1,183 @@
+from enterprise.actions import export_as_csv_action
+
+
+class DummyModel(object):
```
Why not use your real model, `EnterpriseCustomer`?

e-kolpakov replied:
@clintonb For two reasons:

1. To emphasize that the action works with any model
2. `test_export_as_csv_callable_fields` needs a callable attribute on a model, while `EnterpriseCustomer` does not have any properties/methods so far, and I don't want to create one on the actual model just for the test purposes.

clintonb replied:
This makes sense. However, if you have no real models making use of the functionality being tested by `test_export_as_csv_callable_fields`, why is that functionality even included? That's just dead code.

e-kolpakov replied:
@clintonb It [is used](https://github.com/edx/edx-enterprise/pull/1/files#diff-cb0d755b9f3ef2c188945841d6605f1dR38) with `EnterpriseCustomer` - I just don't want to use `EnterpriseCustomer` for testing it for the reasons I've mentioned above

mattdrayer replied:
It doesn't have to be an either/or here -- both models could be tested sequentially if we want to ensure that the workflow executes correctly for both EnterpriseCustomer and another model.

clintonb commented on tests/test_actions.py:
```diff
@@ -0,0 +1,183 @@
+
+    def setUp(self):
+        """
+        Test suite set up method
```
Useless docstring. Remove.

e-kolpakov replied:
@clintonb pydoctest will not allow me to do so

clintonb commented on tests/test_actions.py:
```diff
@@ -0,0 +1,183 @@
+# -*- coding: utf-8 -*-
```
1. Add docstrings to the test methods explaining what's being tested.
2. Add assertions to actually validate the CSV. Knowing the method was called correctly is not as useful as knowing the method's output is valid. (Too much mocking can be a bad thing.)

clintonb commented on tests/test_models.py:
```diff
@@ -1,27 +1,87 @@
-        pass
+    def test_get_queryset_returns_only_active(self):
+        customer1 = EnterpriseCustomerFactory(active=True)
+        customer2 = EnterpriseCustomerFactory(name="Customer 2", active=True)
```
The names here don't actually matter. Remove them, and update your factories to generate fuzzy text.

clintonb commented on tests/test_models.py:
```diff
@@ -1,27 +1,87 @@
-#!/usr/bin/env python
```
Add a test for `EnterpriseCustomer.all_objects` if you choose to continue filtering by default.

e-kolpakov commented on enterprise/actions.py:
```diff
@@ -0,0 +1,59 @@
+                    value = field()
+                else:
+                    value = field
+                if value is None:
```
@clintonb 

* As you pointed out, empty string would denote both empty string and `None`.
* Some people (including me) prefer viewing CSV in ordinary text editors rather than Microsoft Excel/Libre Office Calc. Empty/null data is harder to read in that case.

These two are weak arguments, but I suppose some extra care for the users justify four lines of code. Unless you've got a stronger argument against it/

e-kolpakov commented on enterprise/admin.py:
```diff
@@ -0,0 +1,110 @@
+# -*- coding: utf-8 -*-
+"""
+Django admin integration for enterprise app.
```
@clintonb as much as I dislike non-informative docstrings myself, one of the quality checks _demand_ virtually everything to have a docstring. So if I just remove it, the build will fail. I'll try to update docstrings to carry at least some useful information, but in some cases there's no better docstring than tautology (i.e. test `setUp` methods).

clintonb replied:
If our tools (pydocstyle, in this case) are forcing us to do the wrong thing (write useless docstrings, in this case), we either need to tweak those tools are abandon them.

e-kolpakov replied:
@clintonb [ok](https://github.com/edx/edx-enterprise/pull/1/commits/bdfb86b777f904745aaf181789deca3ce9d8f4ab)

e-kolpakov commented on enterprise/admin.py:
```diff
@@ -0,0 +1,110 @@
+@admin.register(EnterpriseCustomer)
+class EnterpriseCustomerAdmin(admin.ModelAdmin):
+    """
+    Django admin model for EnterpriseCustomer.
```
https://github.com/edx/edx-enterprise/pull/1#discussion_r84255300

e-kolpakov commented on enterprise/admin.py:
```diff
@@ -0,0 +1,110 @@
+
+    class Meta(object):
+        """
+        Meta class for EnterpriseCustomer admin model.
```
https://github.com/edx/edx-enterprise/pull/1#discussion_r84255300

e-kolpakov commented on enterprise/admin.py:
```diff
@@ -0,0 +1,110 @@
+@admin.register(EnterpriseCustomerHistory)
+class EnterpriseCustomerHistoryAdmin(admin.ModelAdmin):
+    """
+    Django admin model for EnterpriseCustomerHistory.
```
https://github.com/edx/edx-enterprise/pull/1#discussion_r84255300

e-kolpakov commented on enterprise/admin.py:
```diff
@@ -0,0 +1,110 @@
+
+    class Meta(object):
+        """
+        Meta class for EnterpriseCustomerHistory admin model.
```
I would assume you've got the trend from previous three responses. :)

e-kolpakov commented:
@edx/analytics ping - @antoviaque just told me you're interested in reviewing this PR too - any chance you could do this today? I would love to merge this by the end of week, which gives us just one "review-address comments" cycle.

mattdrayer commented on enterprise/models.py:
```diff
@@ -2,42 +2,115 @@
+        name (CharField): Enterprise Customer name
+        short_code (AutoSlugField): Enterprise Customer code - used to reference this Enterprise Customer in other
+                                    parts of the system (SSO, ecommerce, analytics etc.)
+        logo (ImageField):  Enterprise Customer logo
```
`logo` went away...

e-kolpakov replied:
@mattdrayer I'm starting to think that putting fields data into class docstring is not that great idea after all...

mattdrayer commented on tests/test_models.py:
```diff
@@ -1,27 +1,84 @@
+class TestEnterpriseCustomerUser(unittest.TestCase):
     """
-    Tests of the EnterpriseCustomerHistory model.
+    Tests of the EnterpriseCustomerMember model.
```
EnterpriseCustomerMember -> EnterpriseCustomerUser

mattdrayer commented on tests/test_models.py:
```diff
@@ -1,27 +1,84 @@
+        member = EnterpriseCustomerUserFactory(id=customer_user_id, user_id=user_id)
+        expected_to_str = "<EnterpriseCustomerUser {ID}: {customer_name} - {user_id}>".format(
+            ID=customer_user_id,
+            customer_name=member.enterprise_customer.name,
```
member -> user here?

mattdrayer commented on enterprise/admin.py:
```diff
@@ -0,0 +1,68 @@
+    Django admin model for EnterpriseCustomer.
+    """
+
+    list_display = ("name", "short_code", "active", "logo",)
```
remove logo

mattdrayer commented on enterprise/admin.py:
```diff
@@ -0,0 +1,68 @@
+    search_fields = ("name", "short_code",)
+
+    actions = [
+        export_as_csv_action("CSV Export", fields=["name", "active", "short_code", "data_sharing_policy", "logo"])
```
remove logo

mattdrayer commented on enterprise/models.py:
```diff
@@ -2,42 +2,114 @@
-    TODO: replace with a brief description of the model.
+    Enterprise Customer is an organization or a group of people that "consumes" courses.
+
+    Members of Enterprise Customer take courses on the edX platform.
```
We should probably update this docstring to refer to Users instead of Members, although it may need to be slightly rewritten to make good sense

mattdrayer commented:
👍

mulby commented on enterprise/migrations/0001_initial.py:
```diff
@@ -0,0 +1,125 @@
+            field=autoslug.fields.AutoSlugField(blank=True, editable=True,
+                                                help_text='Short code will be used to reference this Enterprise Customer in other parts of the system,i.e. SSO, ecommerce app, analytics etc. It is allowed to override it at creation, but it should never be edited on existing records, as it might break integration with other apps.',
+                                                max_length=120, populate_from='name', unique=True),
+        ),
```
Maybe a naive question - I don't know a whole lot about Django models, but is there a reason this field is altered 5 times after it's defined? Seems like an alteration shouldn't be necessary since it's also defined in this file?!

e-kolpakov replied:
@mulby The plan is to squash all the migrations into one, so in final version there won't be that much of alterations.

mulby commented on enterprise/migrations/0006_auto_20161020_1023.py:
```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
```
Could all of these migrations just be simplified into a single migration?

e-kolpakov replied:
@mulby this is the plan. I've kept them separate for two reasons:

1. It is easier for reviewers to keep track of what's changed after their last review.
2. It is a bit simpler in terms of development process (keeping them squashed to single migration would require migrating back to "zero" migration and then forward again each time something has changed; as opposed to forward-only migration in this case).

mulby commented on requirements/base.in:
```diff
@@ -2,3 +2,8 @@
 django-model-utils        # Provides TimeStampedModel abstract base class
+django-autoslug           # Provides AutoSlugField
+unicodecsv                # Allows exporting CSV with unicode support (a drop-in replacement for built-in csv module)
+Pillow                    # Image manipulation module, required to use ImageField
```
Is this actually used anymore?

e-kolpakov replied:
@mulby Probably not - it was used for `logo` field, which was removed when I was actively trying to wrap this up for the day. Thank you for catching this.

mulby commented on enterprise/models.py:
```diff
@@ -2,42 +2,114 @@
 
-    # TODO: add field definitions
+    enterprise_customer = models.ForeignKey(EnterpriseCustomer, blank=False, null=False)
+    user_id = models.PositiveIntegerField(null=False, blank=False)
```
Assuming this is a foreign key to auth_user (effectively)?

I'm also assuming we wanted to retain the option of not deploying this djangoapp into the same MySQL database as the LMS, so this can't be an explicit foreign key constraint?

e-kolpakov replied:
@mulby Yes, effectively it is a ForeignKey to `User` model. And you're more or less right about motivation behind this - @mattdrayer mentioned plans to move enterprise features into dedicated IDA somewhere 9-12 month from now (if I understood him right), so it makes some sense to be prepared for that.

mulby commented on enterprise/migrations/0001_initial.py:
```diff
@@ -0,0 +1,125 @@
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('user_id', models.IntegerField()),
+                ('enterprise_customer',
```
Perhaps this should be named "enterprise_customer_id"? That would make it more clear that it stores the integer ID (not the short code or the name or some other field).

e-kolpakov replied:
@mulby This is an autogenerated migration, so I trust Django to do what it thinks right.

mulby replied:
I think this comes from the field name, see: https://github.com/edx/edx-enterprise/pull/1/files#diff-4bcae564bc272f5c2e6753dc6021eb1cR93

mulby replied:
Ah, it looks like Django does something smart here. Nevermind!

clintonb commented on enterprise/models.py:
```diff
@@ -2,42 +2,114 @@
+    objects = models.Manager()
+    active_customers = EnterpriseCustomerManager()
+
+    name = models.CharField(max_length=500, blank=False, null=False, help_text=_("Enterprise Customer name."))
```
The comment trail seems to have gone away. We should set a more sane limit here. I understand that some organizations may have lengthy names; but, we must consider how these values will be used. A 500 character name is going to look pretty ridiculous when rendered for display.

clintonb commented on enterprise/models.py:
```diff
@@ -2,42 +2,114 @@
+    active_customers = EnterpriseCustomerManager()
+
+    name = models.CharField(max_length=500, blank=False, null=False, help_text=_("Enterprise Customer name."))
+    short_code = AutoSlugField(
```
This is another comment that seems to have gone away. We should use UUID for values that need to be transmitted across systems. Short code is going to fail when we, inevitably, need to change it because a company changes its name. We can prevent these issues now by using UUID.

Again, 120 characters isn't exactly short.

mattdrayer replied:
we could add a uuid field in addition to short_code, which might be used in more human-oriented scenarios vs. machine-oriented service-to-service interactions.

e-kolpakov replied:
@clintonb short code won't change when the org is renamed. Again, too much discussion for a negligible difference. Will change to UUID and burn all traces of slug field.

clintonb commented on enterprise/models.py:
```diff
@@ -2,42 +2,114 @@
 
-# from django.db import models
+from autoslug import AutoSlugField
+from simple_history.models import HistoricalRecords  # likely a bug in import order checker
```
Bug? Is that why your third-party imports are not sorted alphabetically?

e-kolpakov replied:
@clintonb yes, if I sort them manually `isort --check-only` reports quality violation. If I run `isort` allowing to sort imports it places them in this order (and also adds some newlines).

e-kolpakov commented:
@clintonb I've addressed the remaining comments - if possible could you please take another look early this day (two-three hours from now, the earlier the better) - in that case I'll be able to address small notes (if any) and merge it today (as opposed to Monday).

@mattdrayer Same to you - it would be nice to have an explicit approval of this PR from you.

mattdrayer commented on .gitignore:
```diff
@@ -60,7 +60,7 @@ output/*/index.html
 docs/_build
 docs/modules.rst
-docs/enterprise
+docs/enterprise.rst
```
Starting a formal "Github review" -- will approve shortly

clintonb commented on enterprise/models.py:
```diff
@@ -2,42 +2,108 @@
+    objects = models.Manager()
+    active_customers = EnterpriseCustomerManager()
+
+    short_code = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
```
Field should be named uuid. Using it as a primary key may have a slight performance degradation; but, depending on your scale it may not be enough to matter.
